src/server/bg_services/log_replication_scanner.js

```diff
@@ -0,0 +1,219 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const _ = require('lodash');
+const system_store = require('../system_services/system_store').get_instance();
+const dbg = require('../../util/debug_module')(__filename);
+const system_utils = require('../utils/system_utils');
+const config = require('../../../config');
+const P = require('../../util/promise');
+const Semaphore = require('../../util/semaphore');
+const replication_store = require('../system_services/replication_store').instance();
+const auth_server = require('../common_services/auth_server');
+const cloud_utils = require('../../util/cloud_utils');
+const log_formatter = require('./replication_log_formatter');
+
+class LogReplicationScanner {
+
+    /**
+     * @param {{
+     *   name: string;
+     *   client: nb.APIClient;
+     * }} params
+     */
+    constructor({ name, client }) {
+        this.name = name;
+        this.client = client;
+        this._scanner_sem = new Semaphore(config.REPLICATION_SEMAPHORE_CAP, {
+            timeout: config.REPLICATION_SEMAPHORE_TIMEOUT,
+            timeout_error_code: 'LOG_REPLICATION_ITEM_TIMEOUT',
+            verbose: true
+        });
+        this.noobaa_connection = undefined;
+    }
+
+    async run_batch() {
+        if (!this._can_run()) return;
+        dbg.log0('log_replication_scanner: starting scanning bucket replications');
+        try {
+            if (!this.noobaa_connection) {
+                this.noobaa_connection = cloud_utils.set_noobaa_s3_connection(system_store.data.systems[0]);
+            }
+            await this.scan();
+        } catch (err) {
+            dbg.error('log_replication_scanner:', err, err.stack);
+        }
+        return config.BUCKET_REPLICATOR_DELAY;
+    }
+
+    _can_run() {
+        if (!system_store.is_finished_initial_load) {
+            dbg.log0('log_replication_scanner: system_store did not finish initial load');
+            return false;
+        }
+
+        const system = system_store.data.systems[0];
+        if (!system || system_utils.system_in_maintenance(system._id)) return false;
+
+        return true;
+    }
+
+    async scan() {
+        if (!this.noobaa_connection) throw new Error('noobaa endpoint connection is not started yet...');
+
+        const replications = await replication_store.find_aws_log_based_replication_rules();
+
+        await P.all(_.map(replications, async repl => {
+            _.map(repl.rules, async rule => {
```

Comment:
In my opinion - going through all rules here is incorrect because we can have many rules and the cycle will take lots of time.

Reply:
I'm still worrying about going through all rules here at once,
this function includes list_objects of the logs_bucket, get_object of the next log, parsing, head_object of each candidate * 2 (src & dst buckets) comparison of the results, replication of each object that needs to be replicated